src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -459,40 +464,73 @@ public RestoreInfo getRestoreInfo() {
+
+                if (request.processed) {
+                    return currentState;
+                }
+
+                updatedSnapshotStateQueue.drainTo(drainedRequests);
+                
+                final int batchSize = drainedRequests.size();
+
+                // nothing to process (a previous event has processed it already)
+                if (batchSize == 0) {
+                    return currentState;
+                }
+                
                 MetaData metaData = currentState.metaData();
                 MetaData.Builder mdBuilder = MetaData.builder(currentState.metaData());
                 RestoreMetaData restore = metaData.custom(RestoreMetaData.TYPE);
                 if (restore != null) {
-                    boolean changed = false;
-                    boolean found = false;
-                    ArrayList<RestoreMetaData.Entry> entries = newArrayList();
-                    for (RestoreMetaData.Entry entry : restore.entries()) {
-                        if (entry.snapshotId().equals(request.snapshotId())) {
-                            assert !found;
-                            found = true;
-                            Map<ShardId, ShardRestoreStatus> shards = newHashMap(entry.shards());
-                            logger.trace("[{}] Updating shard [{}] with status [{}]", request.snapshotId(), request.shardId(), request.status().state());
-                            shards.put(request.shardId(), request.status());
-                            if (!completed(shards)) {
-                                entries.add(new RestoreMetaData.Entry(entry.snapshotId(), RestoreMetaData.State.STARTED, entry.indices(), ImmutableMap.copyOf(shards)));
-                            } else {
-                                logger.info("restore [{}] is done", request.snapshotId());
-                                restoreInfo = new RestoreInfo(entry.snapshotId().getSnapshot(), entry.indices(), shards.size(), shards.size() - failedShards(shards));
-                                this.shards = shards;
-                            }
-                            changed = true;
-                        } else {
-                            entries.add(entry);
-                        }
+                    int changedCount = 0;
+                    List<RestoreMetaData.Entry> entries = restore.entries();
+                    batchedRestoreInfo = new RestoreInfo[batchSize];
+                    batchedShards = new Map[batchSize];                    
+
+                    for (int i = 0; i < batchSize; i++) {
+                       UpdateIndexShardRestoreStatusRequest updateSnapshotState = drainedRequests.get(i);
+                       updateSnapshotState.processed = true;
+
+                       boolean found = false;
+                       List<RestoreMetaData.Entry> newEntries = newArrayList();
+                       for (RestoreMetaData.Entry entry : entries) {
```

Comment:
Please see my comment in SnapshotService section about checking for completion and rebuilding the RestoreMetaData again and again. I think the same optimization can be applied here.